internal/regex_helpers.go

```diff
@@ -32,6 +38,99 @@ func MatchNamedCaptureGroups(regEx *regexp.Regexp, content string) map[string]st
 	return results
 }
 
+// MatchNamedCaptureGroupsFromReader matches named capture groups from a reader, assuming the pattern fits within
+// 1.5x the reader chunk size (1MB * 1.5).
+func MatchNamedCaptureGroupsFromReader(re *regexp.Regexp, r io.Reader) (map[string]string, error) {
+	results := make(map[string]string)
+	_, err := processReaderInChunks(r, readerChunkSize, matchNamedCaptureGroupsHandler(re, results))
+	if err != nil {
+		return nil, err
+	}
+	if len(results) == 0 {
+		return nil, nil
+	}
+	return results, nil
+}
+
+// MatchAnyFromReader matches any of the provided regular expressions from a reader, assuming the pattern fits within
+// 1.5x the reader chunk size (1MB * 1.5).
+func MatchAnyFromReader(r io.Reader, res ...*regexp.Regexp) (bool, error) {
+	return processReaderInChunks(r, readerChunkSize, matchAnyHandler(res))
+}
+
+func matchNamedCaptureGroupsHandler(re *regexp.Regexp, results map[string]string) func(data []byte) (bool, error) {
+	return func(data []byte) (bool, error) {
+		if match := re.FindSubmatch(data); match != nil {
+			groupNames := re.SubexpNames()
+			for i, name := range groupNames {
+				if i > 0 && name != "" {
+					results[name] = string(match[i])
+				}
+			}
+			return true, nil
+		}
+		return false, nil
+	}
+}
+
+func matchAnyHandler(res []*regexp.Regexp) func(data []byte) (bool, error) {
+	return func(data []byte) (bool, error) {
+		for _, re := range res {
+			if re.Match(data) {
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+}
+
+// processReaderInChunks reads from the provided reader in chunks and calls the provided handler with each chunk + portion of the previous neighboring chunk.
+// Note that we only overlap the last half of the previous chunk with the current chunk to avoid missing matches that span chunk boundaries.
+func processReaderInChunks(r io.Reader, chunkSize int, handler func(data []byte) (bool, error)) (bool, error) {
+	buf := bufio.NewReader(r)
+	halfChunkSize := chunkSize / 2
+	prevChunk := make([]byte, 0, halfChunkSize)
+	currentChunk := make([]byte, chunkSize)
+
+	for {
+		// read the next chunk
+		n, err := buf.Read(currentChunk)
+		if n > 0 {
+			// combine the last half of the previous chunk with the current chunk
+			combined := make([]byte, len(prevChunk)+n)
```

Comment:
Since this PR is about reducing allocations, I would suggest that there is a ring-buffer-like method that could be used to avoid all the secondary allocations could be avoided and this would only require one copy per iteration. Something sorta like this (it's untested and probably wrong):
```
	chunkSize := 10
	half := chunkSize / 2
	bufSize := chunkSize + half
	buf := make([]byte, bufSize)

	lastRead := 0
	for {
		offset := half
		if lastRead < half {
			offset = lastRead
		}
		start := half - offset
		if lastRead > 0 {
			copy(buf[start:], buf[half+offset:half+lastRead])
		}
		n, err := rdr.Read(buf[half:])
		if err != nil {
			break
		}
		// read buf[start:half+n] 
		lastRead = n
	}
```

Reply:
much better! I'm incorporating now 👍